erc20-migration/hardhat.config.js

```diff
@@ -1,42 +1,249 @@
     },
   },
-  defaultNetwork : "hardhat",  //hardhat is fine for local testing - uncomment below to deploy in a true network
-  /*
+  defaultNetwork: process.env.NETWORK,
   networks: {
     basesepolia: {
       url: "https://sepolia.base.org",
-      accounts: {
-        mnemonic: process.env.MNEMONIC
-      }
+      accounts: [process.env.ADMIN_PRIVK]
     },
     horizenl3: {
-      url: "https://horizen-rpc-testnet.appchain.base.org",
-      accounts: {
-        mnemonic: process.env.MNEMONIC
-      }
+      url: process.env.NETWORK_URL,
+      accounts: [process.env.ADMIN_PRIVK]
     }
   }
-  */
+
 };
 
 task("balances", "Prints the wallet balances", async (taskArgs, hre) => {
-  const accounts = await hre.ethers.getSigners();  
+  const accounts = await hre.ethers.getSigners();
   for (var i = 0; i < 5; i++) {
     console.log(accounts[i].address);
-    console.log( ethers.formatEther(await hre.ethers.provider.getBalance(accounts[i].address)));
+    console.log(ethers.formatEther(await hre.ethers.provider.getBalance(accounts[i].address)));
+  }
+});
+
+function updateCumulativeHash(previousHash, address, value) {
+  //the following hashing algorithm produces the same output as the one used in solidity
+  const encoded = web3.eth.abi.encodeParameters(['bytes32', 'address', 'uint256'], [previousHash, address, value])
+  return web3.utils.sha3(encoded, { encoding: 'hex' })
+}
+
+const EON_VAULT_CONTRACT_NAME = "ZTESTBackupVault"
+const ZEN_TOKEN_CONTRACT_NAME = "ZTEST"
+
+task("contractSetup", "To be used just for testing", async (taskArgs, hre) => {
+
+  console.log("Deploying EONVault contract");
+  const admin = (await ethers.getSigners())[0];
+
+  let factory = await hre.ethers.getContractFactory(EON_VAULT_CONTRACT_NAME);
+  let EONVault = await factory.deploy(admin);
+  let res = await EONVault.deploymentTransaction().wait(); // Wait for confirmation
+
+  if (res.status == 0) {
+    console.error("Deploying EONVault contract failed!");
+    exit(-1);
+  }
+  console.log(`EONVault contract deployed at: ${EONVault.target}`);
+
+  console.log("Deploying ZENToken contract");
+
+  const MOCK_ZEND_VAULT_ADDRESS = "0x0000000000000000000000000000000000000000";
+  factory = await hre.ethers.getContractFactory(ZEN_TOKEN_CONTRACT_NAME);
+  ZENToken = await factory.deploy(MOCK_ZEND_VAULT_ADDRESS, await EONVault.getAddress());
+  console.log(`ZENToken contract deployed at: ${ZENToken.target}`);
+  res = await ZENToken.deploymentTransaction().wait(); // Wait for confirmation
+
+  if (res.status == 0) {
+    console.error("Deploying ZENToken contract failed!");
+    exit(-1);
+  }
+  console.log("Set ERC-20 contract reference in the vault contract");
+  res = await EONVault.setERC20(await ZENToken.getAddress());
+
+  if (res.status == 0) {
+    console.error("Setting ERC-20 contract reference in the vault contract failed!");
+    exit(-1);
+  }
+});
+
+task("restoreEON", "Restores EON accounts", async (taskArgs, hre) => {
```

Comment:
let's separate the task in 2 parts:
- calculation of the hash
- execution
(In the migration day the first step will be executed by multiple developers, for double-check)

In the execution part we may add the hash as input parameter - to further check nothing changed in the middle

Reply:
Added a new task for calculating the EON account hash